sdks/java/core/src/main/java/org/apache/beam/sdk/io/Sink.java

```diff
@@ -259,6 +259,14 @@ public abstract void finalize(Iterable<WriteT> writerResults, PipelineOptions op
     public abstract WriteT close() throws Exception;
 
     /**
+     * Aborts writing the bundle. Removes resources used for writing the bundle,
+     * and closes this {@link Writer}.
+     *
+     * <p>Called for each failed bundle.
```

Comment:
We do not guarantee that the `@Teardown` method will be called on a failed bundle (e.g. what happens if the worker crashes?) The only strong guarantee is that no `DoFn` method will be called on an instance after the call to the `@Teardown` method. We will -usually- call `@Teardown` if a bundle fails, but cannot make guarantees to that effect.